omas/machine_mappings/d3d.json

```diff
@@ -263,13 +275,16 @@
  "gas_injection.pipe.:.second_point.z": {
   "PYTHON": "gas_injection_hardware(ods, {pulse})"
  },
- "gas_injection.pipe.:.valve.:": {
+ "gas_injection.pipe.:.valve_indices": {
+  "PYTHON": "gas_injection_hardware(ods, {pulse})"
+ },
+ "gas_injection.valve.:": {
   "PYTHON": "gas_injection_hardware(ods, {pulse})"
  },
- "gas_injection.pipe.:.valve.:.identifier": {
+ "gas_injection.valve.:.identifier": {
```

Comment:
I expect `gas_injection.pipe` entries to be in the d3d.json
Have you run the `test_machine_mapping_functions()` as done in
https://github.com/gafusion/omas/pull/226/files#diff-85adf0c32372e97156d05366b507ff104e0be31ab974b2cb3b5e2a85c7ebd369R1351

That should take care of editing the `d3d.json` consistently for you...

Reply:
That is good to know.  I didn't realize those files were automated.  I ran `test_machine_mapping_functions` and it reordered some things, but didn't make any substantial changes.  I was not able to run it for the coils_nonaxisymmetric or flux loops though.  I ran into these errors when I tried:

```
===================================
coils_non_axisymmetric_current_data
{'pulse': 133221}
===================================
DEBUG (machine): Reading /home/bechtelt/omas/omas//machine_mappings/support_files/d3d/ccomp.dat
DEBUG (machine): Reading /home/bechtelt/omas/omas//machine_mappings/support_files/d3d/icomp.dat
DEBUG (machine): dim_of(ptdata2("'N1COIL'",133221),0) 	OK	 0.004 secs
DEBUG (machine): dim_of(ptdata2("'C79'",133221),0) 	OK	 0.003 secs
DEBUG (machine): dim_of(ptdata2("'C139'",133221),0) 	OK	 0.003 secs
DEBUG (machine): dim_of(ptdata2("'C199'",133221),0) 	OK	 0.003 secs
DEBUG (machine): dim_of(ptdata2("'C259'",133221),0) 	OK	 0.003 secs
DEBUG (machine): dim_of(ptdata2("'C319'",133221),0) 	OK	 0.003 secs
DEBUG (machine): dim_of(ptdata2("'C19'",133221),0) 	OK	 0.003 secs
DEBUG (machine): dim_of(ptdata2("'PCIU30'",133221),0) 	OK	 0.003 secs
DEBUG (machine): dim_of(ptdata2("'PCIU90'",133221),0) 	OK	 0.002 secs
DEBUG (machine): dim_of(ptdata2("'PCIU150'",133221),0) 	OK	 0.002 secs
DEBUG (machine): dim_of(ptdata2("'PCIL30'",133221),0) 	OK	 0.002 secs
DEBUG (machine): dim_of(ptdata2("'PCIL90'",133221),0) 	OK	 0.003 secs
DEBUG (machine): dim_of(ptdata2("'PCIL150'",133221),0) 	OK	 0.002 secs
DEBUG (machine): dim_of(ptdata2("'PCIU210'",133221),0) 	OK	 0.002 secs
DEBUG (machine): dim_of(ptdata2("'PCIU270'",133221),0) 	OK	 0.002 secs
DEBUG (machine): dim_of(ptdata2("'PCIU330'",133221),0) 	OK	 0.002 secs
DEBUG (machine): dim_of(ptdata2("'PCIL210'",133221),0) 	OK	 0.002 secs
DEBUG (machine): dim_of(ptdata2("'PCIL270'",133221),0) 	OK	 0.002 secs
DEBUG (machine): dim_of(ptdata2("'PCIL330'",133221),0) 	OK	 0.002 secs
DEBUG (machine): {'ptdata2("\'N1COIL\'",133221)': 'ptdata2("\'N1COIL\'",133221)', 'ptdata2("\'C79\'",133221)': 'ptdata2("\'C79\'",133221)', 'ptdata2("\'C139\'",133221)': 'ptdata2("\'C139\'",133221)', 'ptdata2("\'C199\'",133221)': 'ptdata2("\'C199\'",133221)', 'ptdata2("\'C259\'",133221)': 'ptdata2("\'C259\'",133221)', 'ptdata2("\'C319\'",133221)': 'ptdata2("\'C319\'",133221)', 'ptdata2("\'C19\'",133221)': 'ptdata2("\'C19\'",133221)', 'ptdata2("\'PCIU30\'",133221)': 'ptdata2("\'PCIU30\'",133221)', 'ptdata2("\'PCIU90\'",133221)': 'ptdata2("\'PCIU90\'",133221)', 'ptdata2("\'PCIU150\'",133221)': 'ptdata2("\'PCIU150\'",133221)', 'ptdata2("\'PCIL30\'",133221)': 'ptdata2("\'PCIL30\'",133221)', 'ptdata2("\'PCIL90\'",133221)': 'ptdata2("\'PCIL90\'",133221)', 'ptdata2("\'PCIL150\'",133221)': 'ptdata2("\'PCIL150\'",133221)', 'ptdata2("\'PCIU210\'",133221)': 'ptdata2("\'PCIU210\'",133221)', 'ptdata2("\'PCIU270\'",133221)': 'ptdata2("\'PCIU270\'",133221)', 'ptdata2("\'PCIU330\'",133221)': 'ptdata2("\'PCIU330\'",133221)', 'ptdata2("\'PCIL210\'",133221)': 'ptdata2("\'PCIL210\'",133221)', 'ptdata2("\'PCIL270\'",133221)': 'ptdata2("\'PCIL270\'",133221)', 'ptdata2("\'PCIL330\'",133221)': 'ptdata2("\'PCIL330\'",133221)'} 	all OK	 0.042 secs
Traceback (most recent call last):
  File "d3d.py", line 1354, in <module>
    test_machine_mapping_functions(__all__, globals(), locals())
  File "/home/bechtelt/omas/omas/omas_machine.py", line 623, in test_machine_mapping_functions
    func(ods, **regression_kw)
  File "/home/bechtelt/omas/omas/omas_machine.py", line 578, in machine_mapping_caller
    out = f(*args, **kwargs)
  File "d3d.py", line 482, in coils_non_axisymmetric_current_data
    data_norm=1.0,
  File "/home/bechtelt/omas/omas/machine_mappings/_common.py", line 234, in fetch_assign
    raise RuntimeError(f'Could not determine time info from {TDI} signals')
RuntimeError: Could not determine time info from ptdata2("'N1COIL'",133221) signals
```
and

```
=====================
magnetics_floops_data
{'pulse': 133221}
=====================
DEBUG (machine): Reading /home/bechtelt/omas/omas//machine_mappings/support_files/d3d/112001/mhdin.dat
DEBUG (machine): dim_of(ptdata2("PSF1A",133221),0) 	OK	 0.050 secs
DEBUG (machine): {'ptdata2("PSF1A",133221)': 'ptdata2("PSF1A",133221)', 'ptdata2("PSF2A",133221)': 'ptdata2("PSF2A",133221)', 'ptdata2("PSF3A",133221)': 'ptdata2("PSF3A",133221)', 'ptdata2("PSF4A",133221)': 'ptdata2("PSF4A",133221)', 'ptdata2("PSF5A",133221)': 'ptdata2("PSF5A",133221)', 'ptdata2("PSF6NA",133221)': 'ptdata2("PSF6NA",133221)', 'ptdata2("PSF7NA",133221)': 'ptdata2("PSF7NA",133221)', 'ptdata2("PSF8A",133221)': 'ptdata2("PSF8A",133221)', 'ptdata2("PSF9A",133221)': 'ptdata2("PSF9A",133221)', 'ptdata2("PSF1B",133221)': 'ptdata2("PSF1B",133221)', 'ptdata2("PSF2B",133221)': 'ptdata2("PSF2B",133221)', 'ptdata2("PSF3B",133221)': 'ptdata2("PSF3B",133221)', 'ptdata2("PSF4B",133221)': 'ptdata2("PSF4B",133221)', 'ptdata2("PSF5B",133221)': 'ptdata2("PSF5B",133221)', 'ptdata2("PSF6NB",133221)': 'ptdata2("PSF6NB",133221)', 'ptdata2("PSF7NB",133221)': 'ptdata2("PSF7NB",133221)', 'ptdata2("PSF8B",133221)': 'ptdata2("PSF8B",133221)', 'ptdata2("PSF9B",133221)': 'ptdata2("PSF9B",133221)', 'ptdata2("PSI11M",133221)': 'ptdata2("PSI11M",133221)', 'ptdata2("PSI12A",133221)': 'ptdata2("PSI12A",133221)', 'ptdata2("PSI23A",133221)': 'ptdata2("PSI23A",133221)', 'ptdata2("PSI34A",133221)': 'ptdata2("PSI34A",133221)', 'ptdata2("PSI45A",133221)': 'ptdata2("PSI45A",133221)', 'ptdata2("PSI58A",133221)': 'ptdata2("PSI58A",133221)', 'ptdata2("PSI9A",133221)': 'ptdata2("PSI9A",133221)', 'ptdata2("PSF7FA",133221)': 'ptdata2("PSF7FA",133221)', 'ptdata2("PSI7A",133221)': 'ptdata2("PSI7A",133221)', 'ptdata2("PSF6FA",133221)': 'ptdata2("PSF6FA",133221)', 'ptdata2("PSI6A",133221)': 'ptdata2("PSI6A",133221)', 'ptdata2("PSI12B",133221)': 'ptdata2("PSI12B",133221)', 'ptdata2("PSI23B",133221)': 'ptdata2("PSI23B",133221)', 'ptdata2("PSI34B",133221)': 'ptdata2("PSI34B",133221)', 'ptdata2("PSI45B",133221)': 'ptdata2("PSI45B",133221)', 'ptdata2("PSI58B",133221)': 'ptdata2("PSI58B",133221)', 'ptdata2("PSI9B",133221)': 'ptdata2("PSI9B",133221)', 'ptdata2("PSF7FB",133221)': 'ptdata2("PSF7FB",133221)', 'ptdata2("PSI7B",133221)': 'ptdata2("PSI7B",133221)', 'ptdata2("PSF6FB",133221)': 'ptdata2("PSF6FB",133221)', 'ptdata2("PSI6B",133221)': 'ptdata2("PSI6B",133221)', 'ptdata2("PSI89FB",133221)': 'ptdata2("PSI89FB",133221)', 'ptdata2("PSI89NB",133221)': 'ptdata2("PSI89NB",133221)', 'ptdata2("PSI1L",133221)': 'ptdata2("PSI1L",133221)', 'ptdata2("PSI2L",133221)': 'ptdata2("PSI2L",133221)', 'ptdata2("PSI3L",133221)': 'ptdata2("PSI3L",133221)'} 	all OK	 0.475 secs
DEBUG (machine): Reading /home/bechtelt/omas/omas//machine_mappings/support_files/d3d/btcomp.dat
DEBUG (machine): data(ptdata("'BTI66M053'",133221)) 	NO	 0.005 secs
Traceback (most recent call last):
  File "/home/bechtelt/omas/omas/omas_machine.py", line 837, in raw
    out_results = MDSplus.Data.data(conn.get(TDI))
  File "/fusion/projects/codes/atom/omfit_v3.x/atom_iris_test/miniconda3_master/lib/python3.7/site-packages/MDSplus/connection.py", line 254, in get
    ans = self.__getAnswer__()
  File "/fusion/projects/codes/atom/omfit_v3.x/atom_iris_test/miniconda3_master/lib/python3.7/site-packages/MDSplus/connection.py", line 132, in __getAnswer__
    raise MdsIpException(str(ans))
MDSplus.connection.MdsIpException: %MDSPLUS-E-Unknown, %TDI-E-ABORT, Program requested abort

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "d3d.py", line 1354, in <module>
    test_machine_mapping_functions(__all__, globals(), locals())
  File "/home/bechtelt/omas/omas/omas_machine.py", line 623, in test_machine_mapping_functions
    func(ods, **regression_kw)
  File "/home/bechtelt/omas/omas/omas_machine.py", line 578, in machine_mapping_caller
    out = f(*args, **kwargs)
  File "d3d.py", line 1216, in magnetics_floops_data
    compsig_data = m.data()
  File "/home/bechtelt/omas/omas/omas_machine.py", line 736, in data
    return self.raw(f'data({self.TDI})')
  File "/home/bechtelt/omas/omas/omas_machine.py", line 847, in raw
    raise _excp.__class__(str(_excp) + '\n' + '\n'.join(txt))
MDSplus.connection.MdsIpException: %MDSPLUS-E-Unknown, %MDSPLUS-E-Unknown, %TDI-E-ABORT, Program requested abort
 - server: atlas.gat.com:8000
 - treename: None
 - pulse: 133221
 - TDI: data(ptdata("'BTI66M053'",133221))
```

These could be related to the particular choice of test shot, but I am not familiar enough with what is going on to be confident in that.

Reply:
@jmcclena Can you help?

Reply:
Arg, there is an invalid loop signal where the .data() array size is different than the .dim_of(0) size. Not sure why the array sizes are different, but it can be solved by adding a check for validity on the compensations to flux loops (and probably magnetic probes too)

Reply:
Added a check with [8e7ff1e](https://github.com/gafusion/omas/pull/226/commits/8e7ff1e0baca03db4e51d2352e1450f41f8acf8a)